lib/engine/round/base.rb

```diff
@@ -39,6 +39,33 @@ def initialize(game, steps, **opts)
           step.setup
           step
         end
+
+        if @game.use_engine_v2 # rubocop:disable Style/GuardClause
+          # hash for quick lookup in either direction, e.g.,
+          # { 'track' => Step::Track, Step::Track => 'track' }
+          #
+          # if there are multiple steps of the same class, all but the first have
+          # a number (starting with 2) appended to their string representation
+          @steps_h = @steps.group_by(&:type).each_with_object({}) do |(step_type, grouped_steps), steps_h|
+            if grouped_steps.one?
+              step = grouped_steps[0]
+              steps_h[step_type] = step
+              steps_h[step] = step_type
+            else
+              grouped_steps.each.with_index do |grouped_step, index|
+                key =
+                  if index.zero?
+                    step_type
+                  else
+                    "#{step_type}#{index + 1}"
+                  end
+                steps_h[key] = grouped_step
+                steps_h[grouped_step] = key
+              end
```

Comment:
This seems over-optimised for a block that's going to get called just once a round. Most of the code on both branches of the `if` statement is repeated.

```suggestion
            grouped_steps.each_with_index do |grouped_step, index|
              key = "#{step_type}#{index.zero? ? '' : index + 1}"
              steps_h[key] = grouped_step
              steps_h[grouped_step] = key
            end
```

I've tested this and it's a few milliseconds faster than your code, mostly because `each_with_index` is more efficient that `each.with_index` (it doesn't need to create an Iterator). The change in the inner block is basically the same speed.